solidity/random-beacon/contracts/libraries/Heartbeat.sol

```diff
@@ -0,0 +1,166 @@
+    ///        all correct.
+    /// @param sortitionPool Sortition pool used by the application performing
+    ///        claim verification.
+    /// @return failedMembers Identifiers of members who failed the heartbeat.
+    function verifyFailureClaim(
+        FailureClaim calldata claim,
+        SortitionPool sortitionPool,
+        uint256 nonce
+    ) internal view returns (uint32[] memory failedMembers) {
+        uint256 groupSize = claim.groupMembers.length;
+        // At least half of the members plus one must vote for the claim.
+        uint256 groupThreshold = (groupSize / 2) + 1;
+
+        // Validate failed members indices. Maximum indices count is equal to
+        // the group size and is not limited deliberately to leave a theoretical
+        // possibility to accuse more members than `groupSize - groupThreshold`.
+        validateMembersIndices(claim.failedMembersIndices, groupSize);
+
+        // Validate signatures array is properly formed and number of
+        // signatures and signers is correct.
+        uint256 signaturesCount = claim.signatures.length / signatureByteSize;
+        require(claim.signatures.length != 0, "No signatures provided");
+        require(
+            claim.signatures.length % signatureByteSize == 0,
+            "Malformed signatures array"
+        );
+        require(
+            signaturesCount == claim.signingMembersIndices.length,
+            "Unexpected signatures count"
+        );
+        require(signaturesCount >= groupThreshold, "Too few signatures");
+        require(signaturesCount <= groupSize, "Too many signatures");
+
+        // Validate signing members indices. Note that `signingMembersIndices`
+        // were already partially validated during signature validation.
+        validateMembersIndices(claim.signingMembersIndices, groupSize);
+
+        // Each signing member needs to sign the hash of packed `groupPubKey`
+        // and `failedMembersIndices` parameters.
+        bytes32 signedMessageHash = keccak256(
+            abi.encodePacked(
+                nonce,
+                claim.groupPubKey,
+                claim.failedMembersIndices
+            )
+        );
+
+        address[] memory groupMembersAddresses = sortitionPool.getIDOperators(
+            claim.groupMembers
+        );
+
+        // Verify each signature.
+        bytes memory checkedSignature;
+        for (uint256 i = 0; i < signaturesCount; i++) {
+            uint256 memberIndex = claim.signingMembersIndices[i];
+            checkedSignature = claim.signatures.slice(
+                signatureByteSize * i,
+                signatureByteSize
+            );
+            address recoveredAddress = signedMessageHash
+                .toEthSignedMessageHash()
+                .recover(checkedSignature);
+
+            require(
+                groupMembersAddresses[memberIndex - 1] == recoveredAddress,
+                "Invalid signature"
+            );
+        }
```

Comment:
Would pulling this code and `validateSignatures` function from `DKGValidator` to a common library be any help here?

Reply:
Not sure if it's worth the additional gas/size overhead. Curious about your point @pdyraga.

Reply:
I am fine the way it is TBH.